### ISSUE TITLE
Race condition in login.tag?

### ISSUE THREAD

**Issue by MarkusPint** [TehShrike/state-router-example#13]
I am trying to follow your example and it seems to me that the opts.login is available in login.tag only due to luck. ie. if in login tag you called opts.login right away in `<script>` it would tell you that this function is undefined.

Related to this: https://github.com/riot/riot/issues/1299

I am trying to find a decent way to inject services (in Angular they would be services, in Riot they are just resource abstractions; let's just say modules) into custom tags. So for example I have a <country-selecter> tag, which is a dropdown that lists all the countries. The countries are in a NPM module. I need a way to make them available in the country-selecter tag's `<script>` section. Or is this an anti-pattern? Sorry for going off-topic, but it's really hard to find any information on RiotJS best practices...

**Comment by TehShrike** [comment]
You are correct that if you called `opts.login()` right away in the tag, it would be undefined, but it is not by luck that it works.

You can see the timeline of a state change in [this bullet-point list in the readme](https://github.com/TehShrike/abstract-state-router/#state-change-flow).

The third-to-last step is to call the `render` function.  In the riot-state-renderer, this [mounts the tag](https://github.com/TehShrike/riot-state-renderer/blob/master/index.js#L26).

The next step is to call the `activate` function.  The login screen's `activate` function defines `opts.login`: https://github.com/TehShrike/state-router-example/blob/gh-pages/implementations/riot/login/login.js#L13

No user interaction should be possible between those two steps, as activating happens immediately before rendering: https://github.com/TehShrike/abstract-state-router/blob/master/index.js#L201

The `activate` function is great for connecting events from the dom API (Riot, in your case) to the rest of your app.  But what you're looking for is a way to compose some data with a tag in your view.

This is exactly what the `resolve` function is great at - your state can define a `resolve` function that returns an object with whatever data you want, like a list of countries.

The `resolve` function works asynchronously, so you can fetch the data from a database as easily as from a local array.

Unlike whatever you set up in your `activate` function, the data returned by your `resolve` function is accessible to your view from the moment it is first created.  See [here](https://github.com/TehShrike/abstract-state-router/blob/master/index.js#L90) where the abstract-state-router passes the resolve data to the render function, and [here](https://github.com/TehShrike/riot-state-renderer/blob/master/index.js#L26), where the riot-state-renderer passes that content to the mount function.

To see this in action in the state-router-example, check out the `topics` state, which [asynchronously fetches the list of topics](https://github.com/TehShrike/state-router-example/blob/gh-pages/implementations/riot/app/topics/topics.js#L13-L17) which is [immediately displayed in the view](https://github.com/TehShrike/state-router-example/blob/gh-pages/implementations/riot/app/topics/topics.tag?ts=3#L6-L9) as soon as it is rendered.

**Comment by MarkusPint** [comment]
Aha! The topics example is spot on. I also found another way of doing this:

```
<country-selecter>
    <script type="es6">
        let CountryData = require( "country-data" );
        this.countries = CountryData.countries.all;
    </script>
</country-selecter>
```

This also fits well with the loosely decoupled components ideology, because depending on the country-data module concerns only the country-selecter tag, not it's parent tags. It's parent tags should not be handling it's dependencies. You see what I am getting at? Do you see anything wrong with this method?

**Comment by TehShrike** [comment]
I can understand not wanting to unnecessarily expose the country data to a larger scope than necessary.

On the other hand, I don't like the view to be responsible for loading dependencies any more than necessary.

At the moment I tend towards loading all data in the resolve function.  It's especially needed for asynchronously loaded data - loading data in the resolve means that your view doesn't have to handle states where some data hasn't been populated yet.

**Comment by TehShrike** [comment]
Have any other thoughts to add @MarkusPint ? If not, I'll probably close this question. :-)

**Comment by MarkusPint** [comment]
Hey, I think not. I am just beginning to learn Riot, React etc. and it turns out that I started from the wrong end - from the views, whereas a real-world app also needs state management, models, routing etc. I started out thinking React/Riot offer all this by themselves. I am probably not the only one with an Angular background to make this mistake.

Thanks for your help though :) We can close this for now.
